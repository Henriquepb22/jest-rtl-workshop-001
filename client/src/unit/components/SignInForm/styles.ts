const signInFormClasses = ['flex', 'flex-col', 'gap-4', 'w-80'].join(' ')

export { signInFormClasses }
