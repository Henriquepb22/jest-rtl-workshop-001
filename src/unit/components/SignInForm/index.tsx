import React, { useState } from 'react'

import TextField from '../TextField'
import Button from '../Button'

const SignInForm = () => {
  const [values, setValues] = useState({ login: '', password: '' })
  const [loading, setLoading] = useState(false)

  const handleInput = (field: string, value: string) =>
    setValues((oldValues) => ({ ...oldValues, [field]: value }))

  const onSubmit = (e: React.FormEvent) => {
    e.preventDefault()
    setLoading(true)
    console.log(values)
  }

  return (
    <form onSubmit={onSubmit} className="flex flex-col w-80">
      <TextField
        label="Usuário"
        name="login"
        placeholder="Insira seu nome de usuário"
        minLength={4}
        onInputChange={(v) => handleInput('login', v)}
        required
      />
      <TextField
        label="Senha"
        name="password"
        type="password"
        placeholder="Insira sua senha"
        minLength={4}
        onInputChange={(v) => handleInput('password', v)}
        required
      />
      <div className="mt-2">
        <Button disabled={loading}>
          {loading ? 'Carregando...' : 'Entrar'}
        </Button>
      </div>
    </form>
  )
}

export default SignInForm
