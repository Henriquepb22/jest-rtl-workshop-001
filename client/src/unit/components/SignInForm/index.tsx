import React, { useState } from 'react'

import { FieldErrors, signInValidate } from 'utils/validations'
import { useUser } from 'unit/hooks/use-user'
import TextField from '../TextField'
import Button from '../Button'

import { signInFormClasses } from './styles'

const SignInForm = () => {
  const { login } = useUser()
  const [values, setValues] = useState({ username: '', password: '' })
  const [loading, setLoading] = useState(false)
  const [fieldErrors, setFieldErrors] = useState<FieldErrors>({})

  const handleInput = (field: string, value: string) =>
    setValues((oldValues) => ({ ...oldValues, [field]: value }))

  const onSubmit = async (e: React.FormEvent) => {
    e.preventDefault()
    setLoading(true)
    const errors = signInValidate(values)

    if (Object.keys(errors).length) {
      setFieldErrors(errors)
      setLoading(false)
      return
    }
    setFieldErrors({})

    await login(values)
  }

  return (
    <form
      aria-label="Formulário de login" // Necessário para getByRole('form', { name: /formulário de login/i }) funcionar
      onSubmit={onSubmit}
      className={signInFormClasses}
    >
      <TextField
        label="Usuário"
        name="username"
        placeholder="Insira seu nome de usuário"
        onInputChange={(v) => handleInput('username', v)}
        error={fieldErrors?.username}
      />
      <TextField
        label="Senha"
        name="password"
        type="password"
        placeholder="Insira sua senha"
        onInputChange={(v) => handleInput('password', v)}
        error={fieldErrors?.password}
      />
      <Button disabled={loading}>{loading ? 'Carregando...' : 'Entrar'}</Button>
    </form>
  )
}

export default SignInForm
