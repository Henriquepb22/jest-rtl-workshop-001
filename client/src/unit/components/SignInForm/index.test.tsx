import { render, screen } from '@testing-library/react'
import SignInForm from './index'

describe('<SignInForm />', () => {
  it('should render component', () => {
    const { container } = render(<SignInForm />)

    expect(container).toMatchSnapshot()
  })

  it('should render the form correctly', () => {
    render(<SignInForm />)

    expect(
      screen.getByRole('form', { name: /formulÃ¡rio de login/i })
    ).toBeInTheDocument()
  })

  // it('should fill and submit the form', () => {

  // })
  it.todo('should return error if field incorrects')
})
