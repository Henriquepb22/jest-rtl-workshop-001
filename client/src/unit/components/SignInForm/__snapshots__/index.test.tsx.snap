// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInForm /> should render component 1`] = `
<div>
  <form
    aria-label="Formulário de login"
    class="flex flex-col gap-4 w-80"
  >
    <div
      class="flex flex-col"
    >
      <label
        class="font-bold"
        for="username"
      >
        Usuário
      </label>
      <input
        class="bg-gray-200 rounded p-2 border-2 border-gray-300 hover:bg-white hover:border-2 hover:border-pink-600 focus:outline-pink-600 focus:bg-white "
        id="username"
        name="username"
        placeholder="Insira seu nome de usuário"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col"
    >
      <label
        class="font-bold"
        for="password"
      >
        Senha
      </label>
      <input
        class="bg-gray-200 rounded p-2 border-2 border-gray-300 hover:bg-white hover:border-2 hover:border-pink-600 focus:outline-pink-600 focus:bg-white "
        id="password"
        name="password"
        placeholder="Insira sua senha"
        type="password"
        value=""
      />
    </div>
    <button
      class="px-4 py-2 text-base bg-pink-600 text-white rounded w-full transition hover:opacity-80 disabled:opacity-70 disabled:cursor-not-allowed"
    >
      <span>
        Entrar
      </span>
    </button>
  </form>
</div>
`;
